ir/base/reachable.go

```diff
@@ -0,0 +1,202 @@
+		return nil, fmt.Errorf("expecting dict or set")
+	}
+	if dok {
+		return reachableDictAssemble(ctx, d, metadata...)
+	}
+	return reachableSetAssemble(ctx, s, metadata...)
+
+}
+
+func reachableDictAssemble(ctx ir.AssemblerContext, d xr.Dict, metadata ...ir.Metadata) (ir.Node, error) {
+	if d.Tag != "reachable" {
+		return nil, fmt.Errorf("expecting tag reachable")
+	}
+
+	u := d.Copy()
+	u.Tag = ""
+	r := xr.Dict{}
+	for _, p := range d.Pairs {
+		// Check if the value is of type string
+		s, ok := p.Value.(xr.String)
+		if !ok {
+			// If not continue
+			continue
+		}
+		// Check if multiaddr and extract addrinfo
+		info, err := peer.AddrInfoFromString(s.Value)
+		if err != nil {
+			// If not continue
+			continue
+		}
+		// If reachable add pair with multiaddr to reachable
+		if reachable := checkIfReachable(ctx.Host, *info); reachable {
+			r.Pairs = append(r.Pairs, p)
+		}
+		// Remove multiaddr from user dict. Reachability already checked.
+		u.Remove(p.Key)
+
+	}
+	// Assemble reachable and user dicts.
+	asm := ir.DictAssembler{}
+	uasm, err := asm.Assemble(ctx, u, metadata...)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't assemble user dict: %s", err)
+	}
+	rasm, err := asm.Assemble(ctx, r, metadata...)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't assemble reachable dict: %s", err)
+	}
+	return &Reachable{
+		Reachable: rasm,
+		User:      uasm,
+	}, nil
+}
+
+func reachableSetAssemble(ctx ir.AssemblerContext, d xr.Set, metadata ...ir.Metadata) (ir.Node, error) {
+	if d.Tag != "reachable" {
+		return nil, fmt.Errorf("expecting tag reachable")
+	}
+
+	u := d.Copy()
+	u.Tag = ""
+	r := xr.Set{}
+	removed := 0
+	for k, p := range d.Elements {
+		// Check if the value is of type string
+		s, ok := p.(xr.String)
+		if !ok {
+			// If not continue
+			continue
+		}
+		// Check if multiaddr and extract addrinfo
+		info, err := peer.AddrInfoFromString(s.Value)
+		if err != nil {
+			// If not continue
+			continue
+		}
+		// If reachable add pair with multiaddr to reachable
+		if reachable := checkIfReachable(ctx.Host, *info); reachable {
+			r.Elements = append(r.Elements, p)
+		}
+		// Remove multiaddr from user set. Reachability already checked.
+		u.Elements = append(u.Elements[:k-removed], u.Elements[k-removed+1:]...)
+		removed++
+	}
+	// Assemble reachable and user dicts.
+	asm := ir.SetAssembler{}
+	uasm, err := asm.Assemble(ctx, u, metadata...)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't assemble user dict: %s", err)
+	}
+	rasm, err := asm.Assemble(ctx, r, metadata...)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't assemble reachable dict: %s", err)
+	}
+	return &Reachable{
+		Reachable: rasm,
+		User:      uasm,
+	}, nil
+}
+
+// CheckIfReachable Checks if peer reachable with 5s timeout.
+// NOTE: We currently consider a peer reachable if we can dial them.
+// We could change the concept of reachable and say that someone is
+// reachable if we are currently connected to it (host.Network().Peers())
```

Comment:
@aschmahmann What is the current DHT approach? I imagine it is to check whether we are connected to them?

@adlrocha  We should probably have two tags named "connected" and "dialable". This implementation currently corresponds to dialable. Note that you could implement both "connected" and "dialable" in the same Go semantic type (say "Reachable", as you have it now), using one assembler. Just add boolean fields "dialable" and "connected" to Reachable. And in the assembler, make a small adjustment to switch on the tag label.